### DISCUSSION TITLE
Implementing Serilog LogContext Items

### DISCUSSION THREAD

**Discussion by drusellers** [quartznet/quartznet#1908] [Q&A]
Goal: Enrich logs with custom data based on quarts job map data

I'm looking for a way to implement a common logging context enricher for Serilog. My goal is to add a named item in the MergedJobContext into the Serilog context. I'm not aware of an execution pipeline / middleware concept with Quartz, so I'm thinking that this is going to be done by EITHER an `IJobListener` or with a custom base class. I would rather use a middleware / IJobListener situation than a base class because then I don't have to remember to use the base class. Ok, with that said I have 

1. JobListeners are "singletons" correct? 
https://www.quartz-scheduler.net/documentation/quartz-3.x/tutorial/trigger-and-job-listeners.html#using-your-own-listeners

I'm using the IoC version, but I'm pretty sure this is still just a singleton in the end.
```
q.AddJobListener<LogContextJobListener>();
```

Because its a singleton, I can't easily "remove" the log context when its complete. https://github.com/serilog/serilog/wiki/Enrichment

2. There is no "middleware" concept right?

I'm not aware of one, but I would dig it if it was there.

---

Happy to update the docs with any cool stuff that comes from this.

**Comment by lahma** [comment]
> Goal: Enrich logs with custom data based on quarts job map data

And a very good goal it is!

> 1. JobListeners are "singletons" correct?
>    [quartz-scheduler.net/documentation/quartz-3.x/tutorial/trigger-and-job-listeners.html#using-your-own-listeners](https://www.quartz-scheduler.net/documentation/quartz-3.x/tutorial/trigger-and-job-listeners.html#using-your-own-listeners)

That sums it up quite well. They are global and expected to be non-stateful.

> 2. There is no "middleware" concept right?
> 
> I'm not aware of one, but I would dig it if it was there.
> 
> Happy to update the docs with any cool stuff that comes from this.

No middleware, this is an **old** library, but I can see the benefits of having such. Maybe it could even be morphed towards such setup, at least partially and maybe even in non-breaking manner.

I think something similar has been done via diagnostics/OpenTelemetry, maybe this setup shows the problem at its core as it wasn't easy to implement by using existing extensibility constructs like job listeners.

https://github.com/quartznet/quartznet/blob/3e2a857fd4b7791ee8d3ac1dd93141b58141466c/src/Quartz/Core/JobRunShell.cs#L199-L236

I'm open to ideas here, extensibility point (chain/middleware instead of hard-coded single listener/writer), could OT somehow log the things you need?
